File: mdix-java/src/main/test/java/com/midmanstudio/dixscript/HotReloadTest.java
package com.midmanstudio.dixscript;

import org.junit.jupiter.api.*;
import org.junit.jupiter.api.io.TempDir;
import static org.assertj.core.api.Assertions.*;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.FileTime;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;

/**
 * Integration tests for {@link HotReload}.
 * Requires the native lib to be on java.library.path (set by build.gradle.kts).
 * <p>
 * Modified-time changes are set explicitly via {@link Files#setLastModifiedTime} rather
 * than relying on real-time sleeps between writes — several filesystems only offer
 * one-second mtime granularity, which would make a sleep-based test both slow and flaky.
 */
class HotReloadTest {

    private static void write(Path path, String content, Instant mtime) throws IOException {
        Files.write(path, content.getBytes(StandardCharsets.UTF_8));
        Files.setLastModifiedTime(path, FileTime.from(mtime));
    }

    @Test void hasChanged_true_beforeFirstReload(@TempDir Path tmp) throws IOException {
        Path file = tmp.resolve("config.mdix");
        write(file, "@DATA( port = 8080 )", Instant.now());

        try (HotReload watcher = new HotReload(file.toString())) {
            assertThat(watcher.hasLoaded()).isFalse();
            assertThat(watcher.hasChanged()).isTrue();
        }
    }

    @Test void checkAndReload_firstCall_reloadsAndReturnsData(@TempDir Path tmp) throws IOException {
        Path file = tmp.resolve("config.mdix");
        write(file, "@DATA( port = 8080 )", Instant.now());

        try (HotReload watcher = new HotReload(file.toString())) {
            var result = watcher.checkAndReload();
            assertThat(result).isPresent();
            try (Database db = result.get()) {
                assertThat(db.getInt("port")).isEqualTo(8080);
            }
            assertThat(watcher.hasLoaded()).isTrue();
        }
    }

    @Test void checkAndReload_noChange_returnsEmpty(@TempDir Path tmp) throws IOException {
        Path file = tmp.resolve("config.mdix");
        Instant t0 = Instant.now();
        write(file, "@DATA( port = 8080 )", t0);

        try (HotReload watcher = new HotReload(file.toString())) {
            watcher.checkAndReload().ifPresent(Database::close);
            assertThat(watcher.hasChanged()).isFalse();
            assertThat(watcher.checkAndReload()).isEmpty();
        }
    }

    @Test void checkAndReload_afterModification_reloadsAgain(@TempDir Path tmp) throws IOException {
        Path file = tmp.resolve("config.mdix");
        Instant t0 = Instant.now();
        write(file, "@DATA( port = 8080 )", t0);

        try (HotReload watcher = new HotReload(file.toString())) {
            watcher.checkAndReload().ifPresent(Database::close);

            write(file, "@DATA( port = 9090 )", t0.plusSeconds(5));

            var result = watcher.checkAndReload();
            assertThat(result).isPresent();
            try (Database db = result.get()) {
                assertThat(db.getInt("port")).isEqualTo(9090);
            }
        }
    }

    @Test void forceReload_reloadsRegardlessOfChange(@TempDir Path tmp) throws IOException {
        Path file = tmp.resolve("config.mdix");
        write(file, "@DATA( port = 8080 )", Instant.now());

        try (HotReload watcher = new HotReload(file.toString())) {
            watcher.checkAndReload().ifPresent(Database::close);
            try (Database db = watcher.forceReload()) {
                assertThat(db.getInt("port")).isEqualTo(8080);
            }
        }
    }

    @Test void path_returnsConstructorPath(@TempDir Path tmp) throws IOException {
        Path file = tmp.resolve("config.mdix");
        write(file, "@DATA( port = 8080 )", Instant.now());

        try (HotReload watcher = new HotReload(file.toString())) {
            assertThat(watcher.path()).isEqualTo(file.toString());
        }
    }

    @Test void malformedFile_reloadFails_andRetriesOnNextCall(@TempDir Path tmp) throws IOException {
        Path file = tmp.resolve("config.mdix");
        Instant t0 = Instant.now();
        write(file, "@@@INVALID$$$", t0);

        try (HotReload watcher = new HotReload(file.toString())) {
            assertThatThrownBy(watcher::checkAndReload).isInstanceOf(MdixException.class);
            // Failure must not have consumed the "changed" state — a fix-and-retry should still see a change.
            write(file, "@DATA( port = 1 )", t0.plusSeconds(5));
            assertThat(watcher.checkAndReload()).isPresent();
        }
    }

    @Test void missingFile_hasChanged_throws(@TempDir Path tmp) {
        Path file = tmp.resolve("does-not-exist.mdix");
        try (HotReload watcher = new HotReload(file.toString())) {
            assertThatThrownBy(watcher::hasChanged).isInstanceOf(MdixException.class);
        }
    }

    @Test void closedWatcher_throwsOnFurtherUse(@TempDir Path tmp) throws IOException {
        Path file = tmp.resolve("config.mdix");
        write(file, "@DATA( port = 8080 )", Instant.now());

        HotReload watcher = new HotReload(file.toString());
        watcher.close();
        assertThatThrownBy(watcher::hasChanged).isInstanceOf(MdixException.class);
    }

    @Test void close_isIdempotent(@TempDir Path tmp) throws IOException {
        Path file = tmp.resolve("config.mdix");
        write(file, "@DATA( port = 8080 )", Instant.now());

        HotReload watcher = new HotReload(file.toString());
        assertThatCode(() -> { watcher.close(); watcher.close(); }).doesNotThrowAnyException();
    }
}

File: mdix-php/tests/HotReloadTest.php
<?php
declare(strict_types=1);

namespace MidManStudio\Mdix\Tests;

use MidManStudio\Mdix\MdixError;
use MidManStudio\Mdix\MdixHotReload;
use PHPUnit\Framework\TestCase;

/**
 * Integration tests for MdixHotReload.
 * Requires the native lib to be on MDIX_LIB_PATH or in mdix-php/lib/.
 *
 * Modified-time changes are set explicitly via touch() rather than relying
 * on real-time sleeps between writes -- several filesystems only offer
 * one-second mtime granularity, which would make a sleep-based test both
 * slow and flaky.
 */
final class HotReloadTest extends TestCase
{
    private string $tmpFile;

    protected function setUp(): void
    {
        $this->tmpFile = \sys_get_temp_dir() . '/mdix_hotreload_test_' . \bin2hex(\random_bytes(8)) . '.mdix';
    }

    protected function tearDown(): void
    {
        if (\file_exists($this->tmpFile)) {
            \unlink($this->tmpFile);
        }
    }

    private function write(string $content, int $mtime): void
    {
        \file_put_contents($this->tmpFile, $content);
        \touch($this->tmpFile, $mtime);
    }

    public function testHasChangedTrueBeforeFirstReload(): void
    {
        $this->write('@DATA( port = 8080 )', \time());

        $watcher = new MdixHotReload($this->tmpFile);
        try {
            self::assertFalse($watcher->hasLoaded());
            self::assertTrue($watcher->hasChanged());
        } finally {
            $watcher->close();
        }
    }

    public function testCheckAndReloadFirstCallReloadsAndReturnsData(): void
    {
        $this->write('@DATA( port = 8080 )', \time());

        $watcher = new MdixHotReload($this->tmpFile);
        try {
            $db = $watcher->checkAndReload();
            self::assertNotNull($db);
            try {
                self::assertSame(8080, $db->getInt('port'));
            } finally {
                $db->close();
            }
            self::assertTrue($watcher->hasLoaded());
        } finally {
            $watcher->close();
        }
    }

    public function testCheckAndReloadNoChangeReturnsNull(): void
    {
        $t0 = \time();
        $this->write('@DATA( port = 8080 )', $t0);

        $watcher = new MdixHotReload($this->tmpFile);
        try {
            $watcher->checkAndReload()?->close();
            self::assertFalse($watcher->hasChanged());
            self::assertNull($watcher->checkAndReload());
        } finally {
            $watcher->close();
        }
    }

    public function testCheckAndReloadAfterModificationReloadsAgain(): void
    {
        $t0 = \time();
        $this->write('@DATA( port = 8080 )', $t0);

        $watcher = new MdixHotReload($this->tmpFile);
        try {
            $watcher->checkAndReload()?->close();

            $this->write('@DATA( port = 9090 )', $t0 + 5);

            $db = $watcher->checkAndReload();
            self::assertNotNull($db);
            try {
                self::assertSame(9090, $db->getInt('port'));
            } finally {
                $db->close();
            }
        } finally {
            $watcher->close();
        }
    }

    public function testForceReloadReloadsRegardlessOfChange(): void
    {
        $this->write('@DATA( port = 8080 )', \time());

        $watcher = new MdixHotReload($this->tmpFile);
        try {
            $watcher->checkAndReload()?->close();
            $db = $watcher->forceReload();
            try {
                self::assertSame(8080, $db->getInt('port'));
            } finally {
                $db->close();
            }
        } finally {
            $watcher->close();
        }
    }

    public function testPathReturnsConstructorPath(): void
    {
        $this->write('@DATA( port = 8080 )', \time());

        $watcher = new MdixHotReload($this->tmpFile);
        try {
            self::assertSame($this->tmpFile, $watcher->path());
        } finally {
            $watcher->close();
        }
    }

    public function testMalformedFileReloadFailsAndRetriesOnNextCall(): void
    {
        $t0 = \time();
        $this->write('@@@INVALID$$$', $t0);

        $watcher = new MdixHotReload($this->tmpFile);
        try {
            $threw = false;
            try {
                $watcher->checkAndReload();
            } catch (MdixError) {
                $threw = true;
            }
            self::assertTrue($threw);

            // Failure must not have consumed the "changed" state -- a fix-and-retry should still see a change.
            $this->write('@DATA( port = 1 )', $t0 + 5);
            self::assertNotNull($watcher->checkAndReload());
        } finally {
            $watcher->close();
        }
    }

    public function testMissingFileHasChangedThrows(): void
    {
        $watcher = new MdixHotReload(\sys_get_temp_dir() . '/does_not_exist_xyz.mdix');
        try {
            $this->expectException(MdixError::class);
            $watcher->hasChanged();
        } finally {
            $watcher->close();
        }
    }

    public function testClosedWatcherThrowsOnFurtherUse(): void
    {
        $this->write('@DATA( port = 8080 )', \time());

        $watcher = new MdixHotReload($this->tmpFile);
        $watcher->close();

        $this->expectException(MdixError::class);
        $watcher->hasChanged();
    }

    public function testCloseIsIdempotent(): void
    {
        $this->write('@DATA( port = 8080 )', \time());

        $watcher = new MdixHotReload($this->tmpFile);
        $watcher->close();
        $watcher->close();
        $this->addToAssertionCount(1);
    }
}
